### PR TITLE
Middleware Architecture Implementation

### PR DESCRIPTION
**Pull request by muthurathinam**
Changes:
1. Added Support for initializing client with default middleware chain and custom middleware chain
2. Added AuthProvider Implementation for MSAL.js
3. Added and organized unit test cases
4. Reduced output js file size by adding tslib dependency
5. Added documentation for middleware usage, auth provider usage, tasks, content etc
6. Added License File
7. Modified Readme.md file
8. Added License detail in all files
9. Added jsDoc comment in all source files
10. Removed unnecessary public methods buildFullURL, parserPath in GraphRequest and changed some members in LargeFileUploadTask from public to protected
11. Added new rich samples interface for browser
12. Updated node samples

Fixes Bugs:
1. https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/129 
2. https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/128
3. https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/113
4. https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/112
5. https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/42

NOTE: Failure of Travis-ci is because of dependency of a UT is in ES6 which is not supported by older versions of Node (Refer: [Travis Error](https://travis-ci.org/microsoftgraph/msgraph-sdk-javascript/jobs/479373897#L816))

Reviewers Note: Please ignore the changes in files under **lib** folder, because they are generated ones.

### CODE REVIEW

File: docs/CustomAuthenticationProvider.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import { AuthenticationProvider } from "@microsoft/microsoft-graph-client";
+
+class MyAuthenticationProvider implements AuthenticationProvider {
+
+    /**
+     * This method will get called before every request to the msgraph server
+     * This should return a Promise that resolves to an accessToken (in case of success) or rejects with error (in case of failure)
+     * Basically this method will contain the implementation for getting and refreshing accessTokens
+     */
+    public async getAccessToken(): Promise<any> {
```
Comment (darrelmiller) [inline docs/CustomAuthenticationProvider.md]: I'm not sure this interface is flexible enough.  I think we need to pass an array of scopes to enable incremental consent and pass some kind of user identifier in case we are in a multi-user application.

File: docs/GettingRawResponse.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+The graph request will be returned for all the method calls except for actions, because they makes call to the server, so have to save the copy of graph request instance before calling actions
+
+```typescript
+const graphRequest = client.api("/me").select("displayName")
+let response = await graphRequest.get();
+```
+
+**Get the raw response**
+
+Use `.getRawResponse()` method to get the raw response
```
Comment (darrelmiller) [inline docs/GettingRawResponse.md]: What does  this do beyond just call fetch?  Would be possible to just do something like,
```
let rawRequest = graphRequest.GetRawRequest();
let rawResponse = await fetch(rawRequest);
```

File: docs/CreatingClientInstance.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
+# Creating Client Instance
+
+Initialization of the Client can be done in one of below two ways
+
+## 1. Create With ClientOptions [Recommended]
+
+In order to instantiate a Client object, one has to pass in the `authProvider` or `middleware` chain in [ClientOptions](../src/IClientOptions.ts).
+
+### 1. Default Middleware chain
```
Comment (MIchaelMainer) [inline docs/CreatingClientInstance.md]: "1. Default Middleware chain" becomes "Option A. Default Middleware chain". It is confusing to have a numbered heading that matches with the simialrly numbered subheading

File: docs/CustomMiddlewareChain.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+## Middleware
+
+As name suggests it comes in middle of something and that is request and response cycle. It has access to context object which contains request, response objects and any other options that are specific to middleware. Also it has an access to execute next middleware in the chain.
+
+## Step by step procedure
+
+### Implement Middlewares
+
+Create own set of middlewares by implementing [Middleware](../src/IMiddleware.ts) interface. Here two middlewares are created one for handling Logging and another for handling http request and response.
```
Comment (MIchaelMainer) [inline docs/CustomMiddlewareChain.md]: "Create own set of middlewares by implementing Middleware interface. Here two middlewares are created one for handling Logging and another for handling http request and response."

"Create a custom middleware pipeline by implementing the IMiddleware interface. The following examples demonstrate how to create a custom logging middleware and how to create a custom http request a response handler."

File: docs/CustomMiddlewareChain.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+First middleware is passed with the context object containing request, and other middleware specific options. One has to explicitly make call to execute method of the next middleware with context object once the current middleware work is over.
+
+NOTE: Http message handler should set the response object in the context object.
```
Comment (MIchaelMainer) [inline docs/CustomMiddlewareChain.md]: Move this note about the best practice for http message handlers to the line above the example of the http message handler. 

> **Note**: markdown notes are more notable this way.

File: lib/src/IMiddleware.d.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+ * -------------------------------------------------------------------------------------------
+ */
+import { Context } from "./IContext";
+/**
+ * @interface
+ * @property {Function} execute - The method to execute the middleware
+ * @property {Function} [setNext] - A method to set the next middleware in the chain
+ */
+export interface Middleware {
+    execute: (context: Context) => Promise<any>;
```
Comment (MIchaelMainer) [inline lib/src/IMiddleware.d.ts]: Execute should return the results of the call back up promise chain.
`execute: (context: Context) => Promise<any>;`

becomes

`execute: (context: Context) => Promise<Context>;`

File: docs/CustomAuthenticationProvider.md (hunk 1 above)
Comment (muthurathinam) [reply docs/CustomAuthenticationProvider.md]: The example you have seen is a sample implementation of CustomAuthenticationProvider, we have an option to pass an array of scopes to for incremental consent in [MSALAuthenticationProvider](https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/Middleware/src/MSALAuthenticationProvider.ts) with `.addScopes()` method.
For multi user application also we can add similar kind of method to switch the user.

File: docs/GettingRawResponse.md (hunk 2 above)
Comment (muthurathinam) [reply docs/GettingRawResponse.md]: Added ResponseType.RAW which needs to be passed in `.responseType()` get the raw response.
Refer: https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/Middleware/docs/GettingRawResponse.md

File: README.md [hunk 7]
```diff
@@ -1,361 +1,191 @@
...
 
-All calls to Microsoft Graph are chained together starting with `client.api(path)`.  Path supports the following formats:  
+### 2. Authenticate for the Microsoft Graph service
 
-* me
-* /me
-* https://graph.microsoft.com/v1.0/me
-* https://graph.microsoft.com/beta/me
-* me/events?$filter=startswith(subject, 'ship')
+The Microsoft Graph JavaScript Client Library has an adapter implementation ([MSALAuthenticationProvider](src/MSALAuthenticationProvider.ts)) for [MSAL](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-core) (Microsoft Authentication Library) which takes care of getting the `accessToken`. MSAL library does not ship with this library, user have to include it externally (For including MSAL, refer [this](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-core#installation)).
```
Comment (pankypal) [inline README.md]: nit: change "user have to include..." to  "users have to" or "user has to"

File: docs/CreatingClientInstance.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+### Option B. Custom Middleware chain
+
+Want to have complete control over the request and the response objects, one can provide his own chain of middleware.
+Have to pass first middleware in the chain as `middleware` in [ClientOptions](../src/IClientOptions.ts).
+
+```typescript
+let clientOptions: ClientOptions = {
+    // MyFirstMiddleware is the first middleware in my custom middleware chain
+    middleware: new MyFirstMiddleware()
```
Comment (pankajpalOlk) [inline docs/CreatingClientInstance.md]: Can i provide both middleware and authProvider together ?

File: docs/CreatingClientInstance.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+};
+const client = Client.initWithMiddleware(clientOptions);
+```
+
+Refer, [custom middleware chain](./CustomMiddlewareChain.md) for more detailed information.
+
+## 2. Create With Options
+
+Pass an [authProvider function](../src/IAuthProvider.ts) in [Options](../src/IOptions.ts) while initializing the Client. In this case, user have to provide his own implementation for getting and refreshing accessToken. A callback will be passed into this authProvider function, accessToken or error needs to be passed in to that callback.
```
Comment (pankajpalOlk) [inline docs/CreatingClientInstance.md]: nit: 'user has to provide...'

File: docs/CreatingClientInstance.md (hunk 8 above)
Comment (muthurathinam) [reply docs/CreatingClientInstance.md]: No you can't... in such case authProvider will be taken into account.

Comment (muthurathinam) [reply docs/CreatingClientInstance.md]: I guess its better to throw an error in that case. will do that.

File: src/Client.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+            }
+        }
+        return Client.initWithMiddleware(clientOptions);
+    }
+
+    /**
+     * @public
+     * @static
+     * To create a client instance with the Client Options
+     * @param {ClientOptions} options - The options object for initializing the client
```
Comment (pankajpalOlk) [inline src/Client.ts]: nit: change options to clientOptions

File: src/MSALAuthenticationProvider.ts [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    /**
+     * @constructor
+     * Creates an instance of MSALAuthenticationProvider
+     * @param {string} clientId - The clientId value of an application
+     * @param {string[]} scopes - An array of graph scopes
+     * @param {any} [options] - An options object for MSAL initialization
+     * @returns An instance of MSALAuthenticationProvider
+     */
+    constructor(clientId: string, scopes: string[], options?: any) {
+        const callback = (errorDesc, token, error, tokenType) => {
```
Comment (pankajpalOlk) [inline src/MSALAuthenticationProvider.ts]: no default implementation of this callback?

File: src/middleware/HTTPMessageHandler.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+     * @public
+     * @async
+     * To execute the current middleware
+     * @param {Context} context - The request context object
+     * @returns A promise that resolves to Nothing
+     */
+    public async execute(context: Context): Promise<void> {
+        try {
+            context.response = await fetch(context.request, context.options);
+            return;
```
Comment (pankajpalOlk) [inline src/middleware/HTTPMessageHandler.ts]: return statement not required.

File: tsconfig.json [hunk 13]
```diff
@@ -4,22 +4,22 @@
     "noEmitOnError": true,
     "noImplicitAny": false,
     "moduleResolution": "node",
+    "importHelpers": true,
     "removeComments": false,
     "sourceMap": true,
     "target": "es5",
-    "lib": ["es7", "dom"],
+    "lib": ["dom", "es7"],
```
Comment (pankajpalOlk) [inline tsconfig.json]: why this change ?

File: docs/CreatingClientInstance.md (hunk 8 above)
Comment (pankajpalOlk) [reply docs/CreatingClientInstance.md]: also, add this to the documentation..

File: tsconfig.json (hunk 13 above)
Comment (muthurathinam) [reply tsconfig.json]: Just alphabetical ordering.

File: src/middleware/HTTPMessageHandler.ts (hunk 12 above)
Comment (muthurathinam) [reply src/middleware/HTTPMessageHandler.ts]: Its just to bring in the readability that it doesn't return anything.

File: src/MSALAuthenticationProvider.ts (hunk 11 above)
Comment (muthurathinam) [reply src/MSALAuthenticationProvider.ts]: There is no default implementation.

File: src/GraphRequest.ts [hunk 14]
```diff
@@ -1,520 +1,711 @@
...
+            // Remove the following version
+            let endOfVersionStrPos = path.indexOf("/");
+            // Parse out the version
+            self.urlComponents.version = path.substring(0, endOfVersionStrPos);
+            // Strip version from path
+            path = path.substring(endOfVersionStrPos + 1, path.length);
+        }
+
+        // Strip out any leading "/"
+        if (path.charAt(0) === "/") {
```
[the remark this answers is not in the document]
Comment (muthurathinam) [reply src/GraphRequest.ts]: yes it can contain only https://

File: src/GraphErrorHandler.ts [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+     *      }
+     *  }
+     */
+    private static constructErrorFromResponse(error: any, statusCode: number): GraphError {
+        error = error.error;
+        let gError = new GraphError(statusCode);
+        gError.code = error.code;
+        gError.message = error.message;
+        if (error.innerError !== undefined) {
+            gError.requestId = error.innerError["request-id"];
```
[the remark this answers is not in the document]
Comment (muthurathinam) [reply src/GraphErrorHandler.ts]: Have added string | null type to that property so the host application will know the expected value. Here the host application will be the consumption point

File: docs/CustomAuthenticationProvider.md (hunk 1 above)
Comment (darrelmiller) [reply docs/CustomAuthenticationProvider.md]: Ok. Never mind.  There is a default set of scopes that are passed in the constructor.  There is no harm in having the ability to add to that default set of scopes.  If we need to allow the ability to request a token for a smaller set of scopes in the future we can add an optional parameter to getAccessToken that accepts a new array of scopes.
